src/Compilers/Core/Portable/MetadataReference/CachedAssemblySymbolList.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class CachedAssemblySymbolList
+    {
+        private readonly object _guard = new object();
+        private readonly WeakList<IAssemblySymbol> _cachedSymbols = new WeakList<IAssemblySymbol>();
+
+        internal void ForEach<T>(Action<T> action)
```

Comment:
Consider making this `TSymbol, TArg` so that we can make the `ForEach` used in the `ReferenceManager` non-capturing.